### PR TITLE
[R-package] require lgb.Dataset, remove support for passing 'colnames' and 'categorical_feature' for lgb.train() and lgb.cv()

### LINKED ISSUE #6435: [RFC] remove 'categorical_feature' and 'feature_name' parameters in cv() and train()
## Proposal

I'm requesting comment on the following proposal:

* remove keyword argument `categorical_feature` from `cv()` and `train()` in the R and Python packages
* remove keyword argument `feature_name` from `cv()` and `train()` in the Python package
* remove keyword argument `colnames` from `cv()` and `train()` in the R package

And doing all of these only after the packages issuing deprecation warnings for 2-3 releases.

## Summary

Both the R and Python packages expose functions `cv()` (for cross-validation) and `train()` (for regular entire-dataset training). These functions require a LightGBM `Dataset` object.

The `Dataset` object holds attributes `categorical_features` and `feature_names`, and allows setting those via constructor keyword arguments and `set_{attr}()` methods.

Despite that, these `cv()` and `train()` functions *also* take `categorical_features` and `feature_names` as keyword arguments.

Python `cv()`

https://github.com/microsoft/LightGBM/blob/92a8741b9cb36a13d7232272e7052ca5e2fa04e6/python-package/lightgbm/engine.py#L569-L570

Python `train()`

https://github.com/microsoft/LightGBM/blob/92a8741b9cb36a13d7232272e7052ca5e2fa04e6/python-package/lightgbm/engine.py#L62-L63

R-package `cv()`

https://github.com/microsoft/LightGBM/blob/92a8741b9cb36a13d7232272e7052ca5e2fa04e6/R-package/R/lgb.cv.R#L90-L91

R-package `train()`

https://github.com/microsoft/LightGBM/blob/92a8741b9cb36a13d7232272e7052ca5e2fa04e6/R-package/R/lgb.train.R#L57-L58

These keyword arguments aren't providing any value, in my opinion. Their values are just forwarded along to calls like this:

https://github.com/microsoft/LightGBM/blob/92a8741b9cb36a13d7232272e7052ca5e2fa04e6/python-package/lightgbm/engine.py#L738-L740

Which at best is redundant with the `Dataset` class, and at worst could lead to runtime exceptions (if the `Dataset` has already been constructed).

## Motivation

Would simplify the library's interface without any loss of functionality.

If this proposal is accepted, the `Dataset` class would be the only place that this information is provided to `train()` and `cv()`.

## References

Inspired by this post I noticed on Stack Overflow: https://stackoverflow.com/questions/78383840/in-lightgbm-why-do-the-train-and-the-cv-apis-accept-categorical-feature-argument/78405996#78405996

`xgboost` does not expose such arguments in `train()` ([code link](https://github.com/dmlc/xgboost/blob/5e64276a9b95df57e6dd8f9e63347636f4e5d331/python-package/xgboost/training.py#L52)) or `cv()` ([code link](https://github.com/dmlc/xgboost/blob/5e64276a9b95df57e6dd8f9e63347636f4e5d331/python-package/xgboost/training.py#L410)).

These arguments have been part of the API since September 2017: https://github.com/microsoft/LightGBM/commit/ef77806934f6b020c6425372f705fd0240a68f5b#diff-9bd633ead0bdfe9540c42a618efd9e559cca16c522ad844a09fcf4ffc7d6e84c.

### PR DESCRIPTION
**Pull request by jameslamb**
Fixes #6435

Enforces deprecations that have been in `{lightgbm}` sinvce v4.4.0:

* removes support for passing anything other than an `lgb.Dataset` to `data` argument of `lgb.cv()`
* removes `colnames` and `categorical_feature` arguments from `lgb.train()` and `lgb.cv()`

Also fixes some new warnings from `{lintr}` 3.2.0: https://github.com/microsoft/LightGBM/pull/6714#issuecomment-2655040655

### Still need to do reverse dependency checks

~Before this is merged, we should check that it won't break any projects on CRAN that depend on `{lightgbm}`. I might propose introducing a comment-triggered CI job we can use for this purpose, maybe with https://github.com/r-lib/revdepcheck.~

Looks like no reverse dependencies would be broken by the v4.6.0 release!

https://github.com/microsoft/LightGBM/pull/6714#issuecomment-2649842114

### `xgboost` did something similar recently

As @mayer79 mentioned on #6435. See https://github.com/dmlc/xgboost/pull/10031

### CODE REVIEW

**Comment by jameslamb** [comment]
Setting up a CI job to run reverse dependency checks is taking too long (#6734), let's not block this on that.

Tomorrow, I will just test all the reverse dependencies manually. I won't merge this until I've done that and posted a comment here summarizing that.

But opening this up now for review... @StrikerRUS or @jmoralez could you review? This + #5010 are the last 2 things I'd like to try to get into v4.6.0, and as a reminder we only have 5 more days to put up a v4.6.0 release on CRAN (#6791).

Neither of those is **critical**, so if you do not have time to review please just let me know and we can skip them for this release.

**Comment by jameslamb** [comment]
Listing out the reverse dependencies. I'll update this comment as I run the tests (probably tomorrow, my timezone, at this point, sorry).

## Reverse Imports

**`{cbl}`**

**`{misspi}`**

**`{predhy}`**

**`{predhy.GUI}`**

**`{sae.projection}`**

**`{ubair}`**

## Reverse Suggests

**`{bonsai}`**

**`{EIX}`**

**`{fastml}`**

**`{mllrnrs}`**

**`{qeML}`**

**`{r2pmml}`**

**`{SHAPforxgboost}`**

**`{stackgbm}`**

**`{treeshap}`**

## Reverse Enhances

**`{fastshap}`**

**`{shapviz}`**

**`{vip}`**

**Comment by mayer79** [comment]
I like these changes,thanks for working on this.

**Comment by jameslamb** [comment]
Also posting because edits wouldn't have generated a notification.... I've finished reverse dependency checks. Details here: https://github.com/microsoft/LightGBM/pull/6714#issuecomment-2649842114

Happy to say that I don't think v4.6.0 will cause any problems for projects on CRAN depending on `{lightgbm}`!

File: R-package/tests/testthat/test_lgb.Booster.R [hunk 1]
```diff
@@ -955,7 +955,7 @@ test_that("Saving a model with unknown importance type fails", {
 
 .params_from_model_string <- function(model_str) {
     file_lines <- strsplit(model_str, "\n", fixed = TRUE)[[1L]]
-    start_indx <- which(grepl("^parameters\\:$", file_lines)) + 1L
+    start_indx <- which(file_list == "parameters:") + 1L
```
Comment (jameslamb) [inline R-package/tests/testthat/test_lgb.Booster.R:958]:
```suggestion
    start_indx <- which(file_lines == "parameters:") + 1L
```

oops

**Comment by jameslamb** [comment]
Going to merge this on the stale approval to keep the release moving, and because the small set of linter-related changes I made should be non-controversial.

#6796 will need to be approved anyway, so @jmoralez @StrikerRUS if you have any follow-up comments please make them there and we can incorporate them before releasing.

Thanks!
